### ISSUE TITLE
Error when repeatedly appending data to a JLD2 file using groups

### ISSUE THREAD

**Issue by simonp0420** [JuliaIO/JLD2.jl#307]
Sorry for the noise.  My "minimal example" doesn't really exercise the bug.  I will try to create a useful MWE.  Closing issue for now.

**Comment by simonp0420** [comment]
OK, I now have a repeatable minimal example.  I apologize for its length, but I haven't been able to reproduce the problem with anything less. 

Here are the contents of "jld2test.jl":
```Julia
module PSSFSS


module GSMs
export GSM

mutable struct GSM{T1 <: AbstractMatrix, T2 <: AbstractMatrix}
    s11::T1
    s12::T2
    s21::T2
    s22::T1
end

end # module GSMs
##############################################
module Outputs
using ..GSMs: GSM
using JLD2
using StaticArrays
export runit

const SV2 = SVector{2,Float64}

SteerType = Union{NamedTuple{(:ψ₁, :ψ₂), Tuple{Float64, Float64}}, 
                  NamedTuple{(:θ, :ϕ), Tuple{Float64, Float64}}}

struct Result
    gsm::GSM
    steering::SteerType
    β⃗₀₀::SV2
    FGHz::Float64
    ϵᵣin::ComplexF64
    μᵣin::ComplexF64
    β₁in::SV2
    β₂in::SV2
    ϵᵣout::ComplexF64
    μᵣout::ComplexF64
    β₁out::SV2
    β₂out::SV2
end

function append_result_data(fname, gname, result)
    jldopen(fname, "a") do fid
        group = JLD2.Group(fid, gname)
        group["result"] = result
    end
    return    
end

function read_result_file(fname::AbstractString)::Vector{Result}
    dat = load(fname) # a Dict
    ks = collect(keys(dat))
    sort!(ks, by = x -> parse(Int,split(x, '/')[1]))
    Result[dat[k] for k in ks]
end

function runit(n1=1)
    infile = "pssfss.res"
    outfile = "test.res"
    isfile(outfile) && rm(outfile)
    results = read_result_file(infile)
    for k = n1:length(results)
        println(k)
        gname = string(k)
        append_result_data(outfile, gname, results[k])
    end
end

end # module Outputs
##############################################

end # module PSSFSS
```

The JLD2 file used as input above, "pssfss.res", can be downloaded from [here](https://github.com/simonp0420/jld2_test).
I execute this in a clean Julia 1.6.1 environment that only has `JLD2` added by the package manager:

```julia
~/.../packages/jld2_test >>> julia                                                                                                               
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.1 (2021-04-23)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(@v1.6) pkg> activate .
  Activating environment at `/run/timeshift/backup/simonp_win/julia/packages/jld2_test/Project.toml`

(jld2_test) pkg> status
      Status `/run/timeshift/backup/simonp_win/julia/packages/jld2_test/Project.toml`
  [033835bb] JLD2 v0.4.4

julia> include("jld2test.jl")
Main.PSSFSS

julia> PSSFSS.Outputs.runit()
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
ERROR: JLD2.InvalidDataException("")
Stacktrace:
  [1] load_group(f::JLD2.JLDFile{JLD2.MmapIO}, roffset::JLD2.RelOffset)
    @ JLD2 ~/.julia/packages/JLD2/KN6F6/src/groups.jl:322
  [2] jldopen(fname::String, wr::Bool, create::Bool, truncate::Bool, iotype::Type{JLD2.MmapIO}; fallback::Type{IOStream}, compress::Bool, mmaparrays::Bool)
    @ JLD2 ~/.julia/packages/JLD2/KN6F6/src/JLD2.jl:306
  [3] jldopen(fname::String, wr::Bool, create::Bool, truncate::Bool, iotype::Type{JLD2.MmapIO})
    @ JLD2 ~/.julia/packages/JLD2/KN6F6/src/JLD2.jl:249
  [4] jldopen(fname::String, mode::String; iotype::Type, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JLD2 ~/.julia/packages/JLD2/KN6F6/src/JLD2.jl:341
  [5] jldopen(fname::String, mode::String)
    @ JLD2 ~/.julia/packages/JLD2/KN6F6/src/JLD2.jl:336
  [6] jldopen(::Main.PSSFSS.Outputs.var"#1#2"{String, Main.PSSFSS.Outputs.Result}, ::String, ::Vararg{String, N} where N; kws::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JLD2 ~/.julia/packages/JLD2/KN6F6/src/loadsave.jl:2
  [7] jldopen
    @ ~/.julia/packages/JLD2/KN6F6/src/loadsave.jl:2 [inlined]
  [8] append_result_data
    @ /run/timeshift/backup/simonp_win/julia/packages/jld2_test/jld2test.jl:43 [inlined]
  [9] runit(n1::Int64)
    @ Main.PSSFSS.Outputs /run/timeshift/backup/simonp_win/julia/packages/jld2_test/jld2test.jl:65
 [10] runit()
    @ Main.PSSFSS.Outputs /run/timeshift/backup/simonp_win/julia/packages/jld2_test/jld2test.jl:58
 [11] top-level scope
    @ REPL[4]:1

```
Note that the error occurs in `append_result_data` on the 15th iteration in `runit`.  The original error in my real code occurs also on the 15th iteration.  In the real code, there is no reading done from a JLD2 file; it simply is writing out the results of a long calculation on each iteration.  When I tried to set up a simple data generation/appending example, I couldn't get it to error.  I had to use the actual data generated by my application. 

Any help with this would be greatly appreciated.

**Comment by JonasIsensee** [comment]
Hi @simonp0420 !

Thank you for this detailed report.
I was able to reduce the problem to the following:
```
outfile = "test.jld2"
isfile(outfile) && rm(outfile)
for n = 1:N
    jldopen(outfile, "a") do fid
        group = JLD2.Group(fid, string(n))
        group["result"] = 1
    end
end
```

and I implemented a fix in #308 that I will merge once tests pass.

**Comment by simonp0420** [comment]
Awesome!  I dared not imagine that I could submit the bug report before going to bed and then wake to find it solved.  I owe you a beer!  Thanks!!!
